src/sage/algebras/quatalg/quaternion_algebra.py

```diff
@@ -2256,6 +2259,291 @@ def isomorphism_to(self, other, *, conjugator=False):
+        r"""
+        Return a random left ideal of ``self`` of norm `N`.
+
+        INPUT:
+
+        - ``N`` -- (integer) the norm of the ideal returned
+        - ``primitive`` -- (bool, default: ``True``) if set, the output ideal
+          is guaranteed to be primitive
+
+        OUTPUT: a left ideal of self of chosen norm `N`
+
+        .. WARNING::
+
+            The algorithm requires factoring `N`, so for large input the algorithm
+            will be slow.
+
+        EXAMPLES::
+
+            sage: B = QuaternionAlgebra(next_prime(2**50))
+            sage: O = B.maximal_order()
+            sage: N = 2**10 * 3**10 * 5**10
+            sage: assert O.random_left_ideal_of_norm(N).norm() == N
+
+        Verify that the method works when N is not coprime to `i^2` or `j^2`::
+
+            sage: B = QuaternionAlgebra(-7, -13)
+            sage: O = B.maximal_order()
+            sage: N = 3**5 * 7**5
+            sage: O.random_left_ideal_of_norm(N).norm() == N
+            True
+
+        Verify that the method works when N is not coprime to `i^2` or `j^2`::
+
+            sage: B = QuaternionAlgebra(-7, -13)
+            sage: O = B.maximal_order()
+            sage: N = 3**5 * 7**5
+            sage: O.random_left_ideal_of_norm(N).norm() == N
+            True
+
+        TESTS::
+
+            sage: for _ in range(10):
+            ....:     p, q = [random_prime(100, lbound=10) for _ in range(2)]
+            ....:     ep, eq = [choice([-1, 1]) for _ in range(2)]
+            ....:     B = QuaternionAlgebra(ep * p, eq * q)
+            ....:     if not B.is_division_algebra():
+            ....:         continue
+            ....:     O = B.maximal_order()
+            ....:     N = random_prime(100)
+            ....:     try:
+            ....:         assert O.random_left_ideal_of_norm(N).norm() == N
+            ....:     except (NotImplementedError, ValueError, TypeError, ZeroDivisionError):
+            ....:         pass
+        """
+        if self.base_ring() != ZZ:
+            raise NotImplementedError("only implemented for quaternion algebras over ZZ")
+
+        # it seems this check is not strict, as the method still works sometimes
+        if gcd(self.discriminant(), N) != 1:
+            raise ValueError("the norm must be coprime with the discriminant")
+
+        B, (i, j, _) = self.quaternion_algebra().objgens()
+
+        # the method is not properly implemented when B is not a division algebra
+        # TODO: remove this warning
+        if not B.is_division_algebra():
+            import warnings
+            warnings.warn("B is not a division algebra. The method may run into an infinite loop.")
+
+        I = self.unit_ideal()
+        I_last = I
+        O_i = self
+
+        for (ell, e) in factor(N):
+            cnt = 0
+            # repeat until `e` successful iterations
+            while cnt < e:
+                if gcd(ell, 2 * ZZ(i**2 * j**2)) != 1:
```

Comment:
Is it worth computing `2 * ZZ(i**2 * j**2)` once as a variable and then storing it before the loop?